lively.morphic/layout.js

```diff
@@ -1523,51 +1523,98 @@ export class ConstraintLayout extends Layout {
+    let left, top;
+    switch (horizontalPolicy) {
+      case 'fixed':
+      case 'resize':
+        // does not affect position
+        left = config.left;
+        break;
+      case 'scale':
+        left = config.leftProportion / 100 * this.container.width;
+        break;
+      case 'move':
+        left = this.container.width - submorph.width - config.right;
+        break;
+      case 'center':
+        left = config.leftCenterProportion / 100 * this.container.width - submorph.width / 2;
+        break;
     }
-    // also update the extent if the resize policy is not fixed!
-    if (newWidth !== morph.width && horizontallyResized) {
-      if (makeDirty) morph.width = newWidth;
-      else {
-        morph._morphicState.extent = morph.extent.withX(newWidth);
-        updateTransform = true;
-        signal(morph, 'extent', morph.extent);
-      }
+
+    switch (verticalPolicy) {
+      case 'fixed':
+      case 'resize':
+        // does not affect position
+        top = config.top;
+        break;
+      case 'scale':
+        top = config.topProportion / 100 * this.container.height;
+        break;
+      case 'move':
+        top = this.container.height - submorph.height - config.bottom;
+        break;
+      case 'center':
+        top = config.topCenterProportion / 100 * this.container.height - submorph.height / 2;
+        break;
     }
-    if (newHeight !== morph.height && verticallyResized) {
-      if (makeDirty) morph.height = newHeight;
-      else {
-        morph._morphicState.extent = morph.extent.withY(newHeight);
-        updateTransform = true;
-        signal(morph, 'extent', morph.extent); // does not update the halo!
-      }
+
+    return pt(left, top);
+  }
+
+  computeExtentFromConfig (config, horizontalPolicy, verticalPolicy, submorph) {
+    let width, height;
+
+    switch (horizontalPolicy) {
+      case 'move':
+      case 'fixed':
+      case 'center':
+        // does not affect width
+        width = submorph.width;
+        break;
+      case 'resize':
+        width = this.container.width - config.left - config.right;
+        break;
+      case 'scale':
+        width = (1 - (config.leftProportion + config.rightProportion) / 100) * this.container.width;
+        break;
     }
 
-    if (morph.layout && morph.layout.renderViaCSS) {
-      morph.layout.onDomResize(node, morph);
+    switch (verticalPolicy) {
+      case 'move':
+      case 'fixed':
+      case 'center':
+        // does not affect height
+        height = submorph.height;
+        break;
+      case 'resize':
+        height = this.container.height - config.top - config.bottom;
+        break;
+      case 'scale':
+        height = (1 - (config.topProportion + config.bottomProportion) / 100) * this.container.height;
+        break;
     }
 
-    if (updateTransform) {
-      // trigger the halo if needed
-      morph.updateTransform();
+    return pt(width, height);
+  }
+
+  updateSubmorphViaDom (morph, node, makeDirty = false) {
```

Comment:
Maybe we should think about changing this name? As the whole purpose of this commit is to actually not read values from the DOM anymore, no?

Reply:
That is true, I refrained from refactoring since I did not want to overcomplicate the "DOM free" implementation of the layout. That will definitely come once we have all layouts transitioned (including grid).

Reply:
I'll open a ticket for that, so that we do not forget about it!